src/AgentCommerce/SalesChannel/UpdateCartRoute.php

```diff
@@ -51,62 +56,103 @@ public function updateCart(string $token, Request $request, SalesChannelContext
+        $customerData['id'] = $customer->getId();
+        $customerData['shippingAddress']['id'] = $customer->getDefaultShippingAddress()?->getId();
+        $customerData['defaultShippingAddress'] = $customerData['shippingAddress'];
 
-        $body = \json_decode($request->getContent(), true, flags: \JSON_THROW_ON_ERROR);
-        $request->attributes->set(AbstractPaymentMethodHandler::PAYPAL_PAYMENT_ORDER_ID_INPUT_NAME, $body['payment_method']['token'] ?? null);
+        $toDeleteAddress = null;
+        if (isset($customerData['billingAddress'])) {
+            $customerData['billingAddress']['id'] = $customer->getDefaultBillingAddress()?->getId() ?? Uuid::randomHex();
+            $customerData['defaultBillingAddress'] = $customerData['billingAddress'];
+        } elseif ($customer->getDefaultShippingAddressId() !== $customer->getDefaultBillingAddressId()) {
+            $toDeleteAddress = [['id' => $customer->getDefaultBillingAddressId()]];
 
-        $response = $this->createCartRoute->createCart($request, $salesChannelContext);
+            $customerData['defaultBillingAddressId'] = $customer->getDefaultShippingAddressId();
+        }
 
-        if ($response->isSuccessful()) {
-            if ($response->getObject()->getVars()['validation_status'] === PayPalCart::VALIDATION_STATUS__VALID) {
-                $response->getObject()->assign(['validation_status' => 'READY']);
+        unset($customerData['shippingAddress'], $customerData['billingAddress']);
+
+        $this->customerRepository->update([$customerData], $salesChannelContext->getContext());
+
+        if (!empty($toDeleteAddress)) {
+            $this->customerAddressRepository->delete($toDeleteAddress, $salesChannelContext->getContext());
+        }
+
+        return $this->createSalesChannelContext(
+            $salesChannelContext->getToken(),
+            $salesChannelContext->getSalesChannelId(),
+            $salesChannelContext->getContext()
+        );
+    }
+
+    private function changeShippingMethod(PayPalCart $payPalCart, SalesChannelContext $salesChannelContext): SalesChannelContext
+    {
+        $shippingOptions = $payPalCart->getAvailableShippingOptions();
+        if (!$shippingOptions) {
+            return $salesChannelContext;
+        }
+
+        foreach ($shippingOptions as $shippingOption) {
+            if (!$shippingOption->isSelected()) {
+                continue;
             }
 
-            $response->setStatusCode(Response::HTTP_OK);
+            if ($shippingOption->getId() === $salesChannelContext->getShippingMethod()->getId()) {
+                // Right shipping method already selected
+                break;
+            }
+
+            try {
+                $this->contextSwitchRoute->switchContext(new RequestDataBag([SalesChannelContextService::SHIPPING_METHOD_ID => $shippingOption->getId()]), $salesChannelContext);
```

Comment:
Please use the returned token to get the new sales channel context:
```suggestion
                $token = $this->contextSwitchRoute->switchContext(new RequestDataBag([SalesChannelContextService::SHIPPING_METHOD_ID => $shippingOption->getId()]), $salesChannelContext)->getToken();
```